### PR TITLE
Feature - Search2 MIME type filtering support

### PR DESCRIPTION
**Pull request by edalex-ian**
##### Checklist

- [x] the [contributor license agreement][] is signed
- [x] commit message follows [commit guidelines][]
- [x] tests are included
- [x] screenshots are included showing significant UI changes
- [x] documentation is changed or added (apidocs)

##### Description of change

Adds support for filtering by a list of MIME types. We need this for a few things, but my immediate concern is the image gallery.

image

[contributor license agreement]: https://www.apereo.org/node/676
[commit guidelines]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

**Comment by PenghaiZhang** [comment]
I noticed that on your screenshot you did a search with filtering by jpeg but the result had a `mp4` attachment.

File: Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/search/SearchHelper.scala [hunk 1]
```diff
@@ -66,6 +65,8 @@ object SearchHelper {
     search.setUseServerTimeZone(true)
     search.setQuery(params.query)
     search.setOwner(params.owner)
+    search.setMimeTypes(
+      (if (params.mimeTypes.isEmpty) None else Some(params.mimeTypes.toList.asJava)).orNull)
```
Comment (PenghaiZhang) [inline Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/search/SearchHelper.scala:69]: I was wondering if we can simplify above  code to
```scala
  search.setMimeTypes(params.mimeTypes)
```

I remember the reason we made it more complicated was to respect Scala coding pattern - avoid null as much as possible. But I might miss something else.

**Review by PenghaiZhang — APPROVED** [review]
Looks good to me Ian!

**Comment by edalex-ian** [comment]
> I noticed that on your screenshot you did a search with filtering by jpeg but the result had a `mp4` attachment.

Yes, maybe not the best screenshot. That item has both videos and images, and it just happens that the top-most attachment is the video one (unfortunately).

So the way this filter works is it returns all items which have at least one attachment with the specified mime-type. And the search result still just returns the items details plus the list of all its attachments.

**Comment by PenghaiZhang** [comment]
> > I noticed that on your screenshot you did a search with filtering by jpeg but the result had a `mp4` attachment.
> 
> Yes, maybe not the best screenshot. That item has both videos and images, and it just happens that the top-most attachment is the video one (unfortunately).
> 
> So the way this filter works is it returns all items which have at least one attachment with the specified mime-type. And the search result still just returns the items details plus the list of all its attachments.

Oh I see. That's all good.

File: Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/search/SearchHelper.scala (hunk 1 above)
Comment (edalex-ian) [reply Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/search/SearchHelper.scala]: Ah yes, here we definitely don't need that it seems. Thank you, addressed in 61703a7

**Comment by edalex-ian** [comment]
Will merge on :green_circle:

**Comment by edalex-ian** [comment]
Gah! That change breaks the `NewSearchPageTest` - even though it was fine on the `Search2ApiTest`. Will see what's happening - and why it could get past the `Search2ApiTest`.

**Comment by edalex-ian** [comment]
Hmmm, I'm not so sure... Is this test a bit flakey... Will re-run while I play locally.

**Comment by edalex-ian** [comment]
Looks like there was a smidge of flake in the `NewSearchPageTest`, so have tuned that up a bit. Will see if we can get that :green_circle: . :crossed_fingers:

**Pull request merged**